feat(api): add command to configure pipette volume in Protocol Engine

Laura-Danielle opened the pull request:
# Overview

Add the configure for liquid class command to protocol engine and allow virtual configs to return different liquid class configurations.

# Changelog

- Add new command to configure liquid for volume
- Add a new equipment command that will configure the volume in the hardware controller and update the pipette state store
- Update the hardware controller to again have a command to configure liquid for volume (a copy of work @sfoster1 had previously done)

Laura-Danielle committed:
feat(api): add command to configure liquid class in Protocol Engine

Rather than importing the concept of liquid classes to pipette state, we should just add a command
to update the static configs from the user facing python API.

sfoster1 commented on api/src/opentrons/protocol_engine/actions/actions.py:
```diff
@@ -189,6 +189,15 @@ class AddPipetteConfigAction:
+    """Update a pipette's current static config class based on a new liquid class."""
+
+    pipette_id: str
+    serial_number: str
```
the serial number definitely shouldn't be in here - that would make it impossible for the action to exist in a json protocol that won't know the serial number

sfoster1 commented on api/src/opentrons/protocol_engine/actions/actions.py:
```diff
@@ -189,6 +189,15 @@ class AddPipetteConfigAction:
+
+    pipette_id: str
+    serial_number: str
+    config: pipette_data_provider.LoadedStaticPipetteData
```
This also doesn't seem right - the payload here should be the liquid class name and not much else. That will get executed by being passed down to the hardware.

sfoster1 reviewed (changes requested):
I don't think this is quite what we want. We want an action that changes the liquid class by sending over the new name, and is handled by calling the `set_liquid_class` method in hardware and then updating the loaded pipette config from hardware (or calling an equivalent function when simulating).

Laura-Danielle committed:
do not add a liquid class action

Laura-Danielle committed:
add protocol engine command to configure liquid

Laura-Danielle committed:
re-add configure for volume so it can be called from PE

sfoster1 commented on api/src/opentrons/hardware_control/instruments/ot3/pipette.py:
```diff
@@ -136,6 +136,28 @@ def __init__(
+            if volume > (self.liquid_class.max_volume * 0.75):
+                return pip_types.LiquidClasses.default.name
+        if self._liquid_class_name == pip_types.LiquidClasses.default:
+            if volume < (self.liquid_class.max_volume * 0.5):
```
I think this wants to be < than half the max volume of the low volume default

sfoster1 commented on api/src/opentrons/protocol_engine/execution/equipment.py:
```diff
@@ -330,6 +340,106 @@ async def load_module(
+        Returns:
+            A LoadedConfiguredVolumeData object.
+        """
+        # TODO (spp, 2023-05-10): either raise error if using MountType.EXTENSION in
```
copypasta

sfoster1 commented on api/src/opentrons/protocol_engine/execution/equipment.py:
```diff
@@ -330,6 +340,106 @@ async def load_module(
+            except RuntimeError as e:
+                raise FailedToLoadPipetteError(str(e)) from e
+
+            pipette_dict = self._hardware_api.get_attached_instrument(
```
don't need to do this

TamarZanzouri commented on api/src/opentrons/protocol_engine/commands/configure_for_volume.py:
```diff
@@ -0,0 +1,89 @@
+class ConfigureForVolumeResult(BaseModel):
+    """Result data from execution of an ConfigureForVolume command."""
+
+    pipetteId: str = Field(
```
why do we need the pipetteId in the result?

sfoster1 replied:
we don't really, might remove it

sfoster1 commented on api/src/opentrons/protocol_engine/execution/equipment.py:
```diff
@@ -330,6 +340,106 @@ async def load_module(
+            # behavior of protocol_context.load_instrument, and is used here as a
+            # pipette existence check
+            try:
+                await self._hardware_api.cache_instruments(cache_request)
```
don't need this

sfoster1 commented on api/src/opentrons/protocol_engine/execution/equipment.py:
```diff
@@ -330,6 +340,106 @@ async def load_module(
+            # is only to support protocol analysis, since the hardware simulator
+            # does not cache requested virtual instruments. Remove per
+            # https://opentrons.atlassian.net/browse/RLIQ-258
+            other_mount = mount.other_mount()
```
don't need the other mount and cache pipette stuff

sfoster1 commented on api/src/opentrons/protocol_engine/execution/equipment.py:
```diff
@@ -330,6 +340,106 @@ async def load_module(
+            is_p50_flex_pipette = (
+                "p50" in pipette_name_value and "flex" in pipette_name_value
+            )
+            if volume < 5 and is_p50_flex_pipette:
```
we should use the exact same logic - maybe make `liquid_class_for_volume` a free function taking 
- the requested volume
- the current liquid class
and reutrning a liquid class, that can get called both here and in hardware?

sfoster1 reviewed (changes requested):
Looks like exactly what we want, save the copypasta in the command implementation, sharing the logic, and fixing that one hysteresis logic defn

TamarZanzouri commented on api/src/opentrons/protocol_engine/execution/equipment.py:
```diff
@@ -330,6 +340,106 @@ async def load_module(
+        # TODO (spp, 2023-05-10): either raise error if using MountType.EXTENSION in
+        #  load pipettes command, or change the mount type used to be a restricted
+        #  PipetteMountType which has only pipette mounts and not the extension mount.
+        use_virtual_pipettes = self._state_store.config.use_virtual_pipettes
+
+        pipette_name_value = (
+            pipette_name.value
+            if isinstance(pipette_name, PipetteNameType)
+            else pipette_name
+        )
+
+        pipette_id = pipette_id or self._model_utils.generate_id()
+
+        if not use_virtual_pipettes:
+            cache_request = {mount.to_hw_mount(): pipette_name_value}
+
+            # TODO(mc, 2022-12-09): putting the other pipette in the cache request
+            # is only to support protocol analysis, since the hardware simulator
+            # does not cache requested virtual instruments. Remove per
+            # https://opentrons.atlassian.net/browse/RLIQ-258
+            other_mount = mount.other_mount()
+            other_pipette = self._state_store.pipettes.get_by_mount(other_mount)
+            if other_pipette is not None:
+                cache_request[other_mount.to_hw_mount()] = (
+                    other_pipette.pipetteName.value
+                    if isinstance(other_pipette.pipetteName, PipetteNameType)
+                    else other_pipette.pipetteName
```
as we spoke about in the morning call. I guess we can extract this logic to a method bc we are doing the same for the `load_pipette`?

sfoster1 replied:
we don't need to do this logic in here at all I think

sfoster1 committed:
refactor: get_liquid_class_for_volume to shared-data

This way it can be shared between hardware and the virtual pipette data
provider. Went back and forth on where to put this but ultimately this
only depends on direct inputs and values from configuration, so
shared-data it is.

sfoster1 committed:
refactor: implement ConfigureForVolume

Unfortunately, the virtual pipette data provider now must be stateful
because it's emulating something stateful, and we should not just move
the state to state because it would be doubling for unavoidable hardware
state like the position of the pipette, and hardware controller really
wants to run this.

That said, we can now nicely use the exact same logic for the virtual
pipette mirror as the actual hardware control.

Needs new tests but doesn't break old ones.

sfoster1 committed:
format shared-data

sfoster1 committed:
add command to unions and update schema

TamarZanzouri commented on api/src/opentrons/hardware_control/instruments/ot3/pipette_handler.py:
```diff
@@ -899,3 +903,15 @@ def get_pipette(self, mount: OT3Mount) -> Pipette:
+        pip = self.get_pipette(mount)
+        if pip.current_volume > 0:
+            # Switching liquid classes can't happen when there's already liquid
+            return
```
do we want to raise here instead?

sfoster1 replied:
yeah probably

TamarZanzouri reviewed (approved):
thank you for the changes! this looks great to me!

Laura-Danielle committed:
add more test coverage for virtual pip config and new shared data func

sfoster1 committed:
refactor(api): Add PrivateResults to PE commands

When PE loaded pipettes, it would set the static config in its state by
having the load method of the equipment handler side-effect emit a
special action that only existed to update the configuration. This
worked ok as long as you weren't too picky about wanting to be sure that
the action was handled before the result of the command was evaluated.

The problem is that now we do care about that, a lot (though it's not
quite as evident in this commit). When we switch pickup and drop tip to
reset the tip state of a pipette, we're going to need to set the
configuration using this new private result method and rely on the
results in the command results handler.

This is implemented by poking a new typevar into CommandUpdateAction
that can optionally carry a "private result", something that will not be
exposed via HTTP or via the database but can carry information in a
lifetime-synchronous way to the result.

This required a lot of refactoring, but I think it's ended up pretty
cleanly; the only test changes are directly required by the changed code.

TamarZanzouri commented on api/src/opentrons/protocol_engine/commands/command_unions.py:
```diff
@@ -500,3 +514,5 @@
     calibration.MoveToMaintenancePositionResult,
 ]
+
+CommandPrivateResult = Union[None, LoadPipettePrivateResult]
```
are we missing `ConfigureForVolumePrivateResult` ?

sfoster1 replied:
we were!

sfoster1 committed:
load tip return data from tip volume

jbleon95 commented on api/src/opentrons/protocol_engine/commands/command.py:
```diff
@@ -178,3 +180,41 @@ def __init__(
+    ABC,
+    Generic[CommandParamsT, CommandResultT, CommandPrivateResultT],
+):
+    """Abstract command creation and execution implementation.
```
Doc strings for this new Abstract Implementation should be updated to not just be copypasta'd from above.

jbleon95 commented on api/src/opentrons/protocol_engine/commands/configure_for_volume.py:
```diff
@@ -0,0 +1,96 @@
+class ConfigureForVolumeResult(BaseModel):
+    """Result data from execution of an ConfigureForVolume command."""
+
+    pipetteId: str = Field(
```
Is there a reason why were are returning pipetteId in the result, because unless it's changing I don't think it should be included.

Laura-Danielle replied:
All the pipette commands include the pipette id.

SyntaxColoring replied:
Mm, `loadPipette` has a `result.pipetteId` because that's the thing that creates the `pipetteId`. Other pipetting commands, like `aspirate` and `blowOut`, do not. (They instead take a `pipetteId` as input, in `params`.)

I agree with @jbleon95—let's remove `pipetteId` from this result model.

sfoster1 replied:
done

jbleon95 commented on api/src/opentrons/protocol_engine/execution/equipment.py:
```diff
@@ -330,6 +342,52 @@ async def load_module(
+        """Ensure the requested volume can be configured for the given pipette.
+
+        Args:
+            pipette_id: An optional identifier to assign the pipette. If None, an
```
Not optional for this method

TamarZanzouri commented on api/src/opentrons/protocol_engine/state/pipettes.py:
```diff
@@ -181,9 +187,9 @@ def _handle_command(self, command: Command) -> None:  # noqa: ANN101, C901
-                    tip_configuration = static_config.tip_configuration_lookup_table[
-                        static_config.max_volume
-                    ]
+                    tip_configuration = list(
```
do we want to leave a comment about defaulting?

jbleon95 commented on api/src/opentrons/protocol_engine/execution/equipment.py:
```diff
@@ -330,6 +342,52 @@ async def load_module(
+                pipette_id, volume, model
+            )
+
+            serial_number = self._model_utils.generate_id(prefix="fake-serial-number-")
```
the fact that we are regenerating serial number here for the simulated case feels a bit iffy even though in practice I don't believe it will have any actionable impact on how we do analysis. The only place we are using serial number is for `get_calibrated_tip_length` which already is referencing a fake serial number in the first place. Still just calling it out here for maybe a comment or not explaining this.

TamarZanzouri commented:
looks great to me! had a few questions but overall love the private setting change :-)

jbleon95 commented on api/tests/opentrons/protocol_engine/execution/test_equipment_handler.py:
```diff
@@ -69,11 +69,27 @@ def _make_config(use_virtual_modules: bool) -> Config:
     """Mock out move_types.py functions."""
     for name, func in inspect.getmembers(pipette_data_provider, inspect.isfunction):
         monkeypatch.setattr(pipette_data_provider, name, decoy.mock(func=func))
+    monkeypatch.setattr(
```
Since `VirtualPipetteDataProvider` is now a class we should avoid monkeypatching it here and instead make the below pytest fixture a decoy mock of that class, unless there is some functional reason why we are doing it this way.

jbleon95 reviewed (approved):
Reviewed the Protocol Engine portions of this PR, all look good to me outside of minor comments and fixing up the failing CI tests. Really like the private result addition, we can definitely use that in the future to clean up some result cruft.

Laura-Danielle reviewed (commented):
Cannot approve my own pull request but I like the private results update a lot.

sfoster1 committed:
brilliant

sfoster1 committed:
format pipette-definition

sfoster1 committed:
remove pipette_id from cfv result and change comments

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/command.py:
```diff
@@ -25,6 +25,8 @@
 
 CommandResultT = TypeVar("CommandResultT", bound=BaseModel)
 
+CommandPrivateResultT = TypeVar("CommandPrivateResultT", bound=BaseModel)
+
```
Low-priority nitpick: I would not bind this to `BaseModel`, because `BaseModel` is for stuff that's serialized, and private results are deliberately never serialized.

And I guess I'd make `ConfigureForVolumePrivateResult` a `@dataclass` instead of a `BaseModel`.

sfoster1 replied:
that's a good point, there's no reason for it to be like that

sfoster1 replied:
honestly there's not really a reason to have a bound at all in that case tbh

SyntaxColoring commented on api/src/opentrons/protocol_engine/commands/load_pipette.py:
```diff
@@ -46,22 +57,32 @@ class LoadPipetteResult(BaseModel):
         )
 
-        return LoadPipetteResult(pipetteId=loaded_pipette.pipette_id)
+        return LoadPipetteResult(
+            pipetteId=loaded_pipette.pipette_id
+        ), LoadPipettePrivateResult(
+            pipette_id=loaded_pipette.pipette_id,
+            serial_number=loaded_pipette.serial_number,
+            config=loaded_pipette.static_config,
+        )
```
Nitpick for slightly faster analysis:
 
```suggestion
        return LoadPipetteResult.construct(
            pipetteId=loaded_pipette.pipette_id
        ), LoadPipettePrivateResult.construct(
            pipette_id=loaded_pipette.pipette_id,
            serial_number=loaded_pipette.serial_number,
            config=loaded_pipette.static_config,
        )
```

SyntaxColoring reviewed (commented):
Generally looks great to me. Thanks!

sfoster1 committed:
some backups for serial number handling

sfoster1 reviewed (approved):
get rid of my own negative review

sfoster1 committed:
Update api/src/opentrons/protocol_engine/commands/load_pipette.py

Co-authored-by: Max Marrone <max@opentrons.com>

sfoster1 committed:
format, comments

sfoster1 committed:
docstring

sfoster1 committed:
test mock fixups

sfoster1 committed:
remove pydantic from private results
